geopandas/explore.py

```diff
@@ -0,0 +1,900 @@
+from statistics import mean
+from warnings import warn
+
+import geopandas
+from shapely.geometry import LineString
+import numpy as np
+import pandas as pd
+
+_MAP_KWARGS = [
+    "location",
+    "prefer_canvas",
+    "no_touch",
+    "disable_3d",
+    "png_enabled",
+    "zoom_control",
+    "crs",
+    "zoom_start",
+    "left",
+    "top",
+    "position",
+    "min_zoom",
+    "max_zoom",
+    "min_lat",
+    "max_lat",
+    "min_lon",
+    "max_lon",
+    "max_bounds",
+]
+
+
+def _explore(
+    df,
+    column=None,
+    cmap=None,
+    color=None,
+    m=None,
+    tiles="OpenStreetMap",
+    attr=None,
+    tooltip=True,
+    popup=False,
+    highlight=True,
+    categorical=False,
+    legend=True,
+    scheme=None,
+    k=5,
+    vmin=None,
+    vmax=None,
+    width="100%",
+    height="100%",
+    categories=None,
+    classification_kwds=None,
+    control_scale=True,
+    marker_type=None,
+    marker_kwds={},
+    style_kwds={},
+    highlight_kwds={},
+    missing_kwds={},
+    tooltip_kwds={},
+    popup_kwds={},
+    legend_kwds={},
+    **kwargs,
+):
+    """Interactive map based on GeoPandas and folium/leaflet.js
+
+    Generate an interactive leaflet map based on :class:`~geopandas.GeoDataFrame`
+
+    Parameters
+    ----------
+    column : str, np.array, pd.Series (default None)
+        The name of the dataframe column, :class:`numpy.array`,
+        or :class:`pandas.Series` to be plotted. If :class:`numpy.array` or
+        :class:`pandas.Series` are used then it must have same length as dataframe.
+    cmap : str, matplotlib.Colormap, branca.colormap or function (default None)
+        The name of a colormap recognized by ``matplotlib``, a list-like of colors,
+        :class:`matplotlib.colors.Colormap`, a :class:`branca.colormap.ColorMap` or
+        function that returns a named color or hex based on the column
+        value, e.g.::
+
+            def my_colormap(value):  # scalar value defined in 'column'
+                if value > 1:
+                    return "green"
+                return "red"
+
+    color : str, array-like (default None)
+        Named color or a list-like of colors (named or hex).
+    m : folium.Map (default None)
+        Existing map instance on which to draw the plot.
+    tiles : str, xyzservices.TileProvider (default 'OpenStreetMap')
+        Map tileset to use. Can choose from this list of built-in tiles or pass
+        :class:`xyzservices.TileProvider`:
+
+        ``["OpenStreetMap", "Stamen Terrain", “Stamen Toner", “Stamen Watercolor"
+        "CartoDB positron", “CartoDB dark_matter"]``
```

Comment:
Question on this - if we want allow to pass a string name of xyzservices.TileProvider, we need to import it. Are we fine adding it as another dependency?

Reply:
I realised this can be useful elsewhere as well so adding the method to xyzservices in https://github.com/geopandas/xyzservices/pull/93. One that is in, we can use it here to pass string name and get rid of this set of built-in providers.

Reply:
I have added a support for a string name of a xyzservices TileProvider based on `query_name` method. `xyzservices` is treated as an optional library so the check works if it is available and is ignored if it is not.